### DISCUSSION TITLE
CMake variants, Cartesian but only for certain builds

### DISCUSSION THREAD

**Discussion by savvn001** [microsoft/vscode-cmake-tools#2869] [Q&A] [answered]
Hi,

I've been wondering if it's possible to configure the `cmake-variants.yaml` such that, some of the options only apply to certain other options. Or in other words, some options are excluded from others, so the combinations listed aren't a full cartesian product.

Take the example below:

```
buildType:
  default: debug
  description: Select build type
  choices:
    debug:
      short: Debug Firmware
      long: Debug config, logs on and debug info generated
      buildType: DEBUG
    release:
      short: Release Firmware
      long: Relase build, logs off, max optimization
      buildType: RELEASE
    test_host:
      short: Test Host
      long: Config for building unit tests on host machine
      buildType: CTEST
    test_device:
      short: Test Device
      long: Config for building unit tests for target device
      buildType: TEST_DEVICE

CLI_LOGs:
  default: cli_logs
  choices:
    cli_logs:
      short: CLI ON, LOGS ON
      long: CLI ON, logs ON
      settings:
        CLI: true
        LOGS: true
    log_only:
      short: CLI OFF, LOGS ON
      long: CLI OFF, logs ON
      settings:
        CLI: false
        LOGS: true
    none:
      short: CLI OFF, LOGS OFF
      long: CLI OFF, logs OFF
      settings:
        CLI: false
        LOGS: false

SOFTDEVICE:
  default: on
  choices:
    on:
      short: Softdevice ON
      settings:
        SOFTDEVICE: true
    off:
      short: Softdevice OFF
      settings:
        SOFTDEVICE: false

BOARD:
  default: V1
  choices:
    V1:
      short: V1 Board
      settings:
        BOARD: V1
    V2:
      short: V2 Board
      settings:
        BOARD: V2  
```   

In my example, you can see the first option is 4 different build profiles, I switch between them, and it tells CMake which toolchain and optimisation to use etc. Now, the issue is, I want the `CLI_LOGs`, `SOFTDEVICE`, `BOARD` options to **only** be applied to the `debug` and release `build` options. So in essence, the options listed are **not** a full cartesian product.

Is this possible? Or am I thinking about it completely wrong and need to approach it differently? I really like the idea of variants as it allows me to easily flick between different build configurations. But in this particular situation, I need it to not quite be a full matrix combination.

**Comment by bobbrow** [comment, marked as answer]
Unfortunately, there is not a way to set up rules for custom combinations when using variants.  If you'd like to have more control over this. you'd have to switch to using `CMakePresets.json`.  Though CMake presets don't let you specify rules either, you can define presets that consist of a combination of other partial presets such that you could get what you want.

It's on our backlog to come up with a proposal to Kitware that will allow for easier combinatorics with CMake presets.  (e.g. address this: https://gitlab.kitware.com/cmake/cmake/-/issues/22538)

**Comment by savvn001** [comment]
Ah thanks, sure. For now I've just manually made the combinatorics so that it's exactly how I want. It's not scalable but works for now since I don't have a large amount of combinations. I'll probably as a workaround make a simple script to generate a `cmake-variants.yaml` file in the future with the exact combos I need as the project grows.
